doc: update KC docs (#13)
mlabs-haskell/lambda-buffers: docs/compiler.md

# _Compiler_ Spec & Design

The _Compiler_ component sits between the _Frontend_ component and the code
generation component named _Codegen_. The purpose of the _Compiler_ is to
typecheck terms generated by the Frontend (terms referred to as _Compiler Input_)
and perform other additional validation checks. The end goal of the _Compiler_
is to ensure that the _Codegen_ component is capable of processing the _Compiler
Output_.

## _Compiler_ interface

The _Compiler_ operates on the _Compiler Input proto_ - enabling, any _Frontend_
to interface with the _Compiler_ in a language agnostic manner.

Similarly, the _Compiler Output_ is also a _proto_ that is then consumed by
_Codegen_ modules, able to be written in any programming environment capable of
communicating via Google Protocol Buffers.

## Checking type definitions

The primary purpose of the compiler is to typecheck the schemata that users
define via the _Frontend_. The schemata terms are ADT declarations and their
types are kinds.

> Currently the schema language supports:

Author: ```suggestion
Currently the schema language _Compiler_ supports:
```

>  1. types (such as `Int` or `Bool`),

Author: ```suggestion
 1. type terms of kind `Type` (such as `Int` or `Bool`),
```

>  2. and type functions (such as `Maybe` or `Either`).

Reviewer: I'd rather see you talk in terms of Kinds, and say things like,

Currently, the Compiler **accepts** type terms of kind `Type` and `Type -> Type` such as:
1. `Int :: Type`,
2. `Bool :: Type`,
3. `Maybe :: Type -> Type`,
4. `Either :: Type -> Type -> Type`.

All **type variables** are of kind `Type` as well

Author: I mention the all type variables later on.

>  2. and type functions (such as `Maybe` or `Either`).

Author: ```suggestion
 2. and type function terms of kind `Type → Type` (such as `Maybe` or `Either`).
```

There are future plans to expand this to Higher Kinded Types (such as `MaybeT`,
`StateT` etc.) - subject to research into _Codegen_ of such types in the target
languages. The schema language and the _Compiler_ do support recursive
types.

Schemata terms must be monomorphically kinded, with polymorphic kinds defaulting
to monomorphic ones. For example `Phantom a = Phantom` would resolve to the
monomorphic kind `Type → Type` rather than the polymorphic kind `∀a. a → Type`.

<!-- cstml comments: Is this true? I think it's an Aim not a requirement?

> :todo: -->

Reviewer: Correct

In addition to typechecking, the compiler must  perform a special check for
recursive types: It must validate that a recursive type is inhabited (or
inhabitable). The purpose of this check is to ensure that any schema which
passes validation is (in principle) a schema for which type definitions and
typeclass "instances" (which may be simple functions in languages without
typeclass support) can be generated. As an example, the compiler should reject

> types such as `data F a = F (F a)`, which is uninhabited.

Reviewer: Can you add this under a Chapter *Checking type cardinality*

<!-- cstml comments: Is this true? Seems as though this should fail
parsing. :todo: -->

(Provisional:) Finally, the compiler should _normalize_ expressions as far as it

> is able to. For example, it may be possible to define a data type in the schema

Reviewer: I don't think it should be a parse error. You should be able to parens around anything mostly. But this seems like a Frontend concern and it connects with the discussion about `compiler.proto` Ty message and how it currently propagates such constructs all the way to the Compiler. For instance

`data G a = G ((Either) ((Maybe) a) Int)`

Would result in `TyAbs "a" $ TyApp (TyApp Either []) [TyApp (TyApp (TyApp Maybe []) [TyVar a]) [TyRef 'Int']]`

So essentially you have `TyApp SomeTy []` with an empty argument list, which should normalize to `SomeTy`.

If we reformulate the `compiler.proto` Ty to use only binary forms, then this becomes solely a Frontend issue.
